Add the Slice builtins/basics

dario23 opened the pull request.

dario23 committed:
Add the Slice builtins/basics

fpoli commented on prusti-viper/src/encoder/builtin_encoder.rs:
```diff
@@ -133,6 +150,70 @@ impl BuiltinEncoder {
             },
+            BuiltinFunctionKind::SliceLookupPure { slice_elem_ty, return_ty } => {
+                let elem_ty_name = if let vir::Type::TypedRef(ref ty) = slice_elem_ty { ty } else { unreachable!() };
+                let type_pred = format!("Slice${}", elem_ty_name);
```
This does not trigger the encoding of the predicate and will probably generate a consistency error like the one in #489 (the cause might be similar). You should call `Encoder::encode_type_predicate_use`, passing the Rust type of the slice. Either you store that type in the `BuiltinFunctionKind` (adding a `<'tcx>`) or you build a Rust type on the fly with [`TyCtxt::mk_ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.mk_ty).

dario23 replied:
added #499 to pass the rust array type via `PlaceEncoding`, will rebase once that is accepted.

fpoli commented on prusti-viper/src/encoder/builtin_encoder.rs:
```diff
@@ -133,6 +150,70 @@ impl BuiltinEncoder {
+            },
+            BuiltinFunctionKind::SliceLen { slice_elem_ty } => {
+                let elem_ty_name = if let vir::Type::TypedRef(ty_name) = slice_elem_ty { ty_name } else { unreachable!() };
+                let slice_pred = format!("Slice${}", elem_ty_name);
```
Same comment as for `SliceLookupPure`

fpoli commented on prusti-viper/src/encoder/builtin_encoder.rs:
```diff
@@ -133,6 +150,70 @@ impl BuiltinEncoder {
+                let self_var = vir::LocalVar::new("self", vir::Type::TypedRef(slice_pred.clone()));
+
+                vir::Function {
+                    name: format!("Slice${}$len", elem_ty_name),
```
I think this is `fn_name`.

fpoli commented on prusti-viper/src/encoder/builtin_encoder.rs:
```diff
@@ -133,6 +150,70 @@ impl BuiltinEncoder {
+                            vir::PermAmount::Read,
+                        ),
+                    ],
+                    posts: vec![],
```
I think for some programs you'll need to add `result >= 0` as postcondition. You can try to test this with a program that contains `assert!(my_slice.len() >= 0);` or something similar.

dario23 replied:
that won't work yet, the `core::slice::<impl [i32]>::len` (that's how it's formatted in the MIR) function call doesn't know yet that its result is the same as `Slice$i32$len` -- i don't know how to `#[extern_spec]` that because a) not sure how to specify the module for that `<impl [i32]>`, just ignoring it doesn't work; and b) i have no way to refer to `Slice$i32$len` in `extern_spec` i.e. from rust.

i'll attempt to special-case it in `procedure_encoder` the way `panic`, `Box::new` et al. are.

dario23 replied:
did the special-casing attempt, see last commit.

fpoli commented on prusti-viper/src/encoder/mir_encoder/mod.rs:
```diff
@@ -218,26 +218,40 @@ pub trait PlaceEncoder<'v, 'tcx: 'v> {
+                            None,
+                        )
+                    },
+                    _ => unreachable!("index on unsupported type '{:?}'", base_ty),
```
Should this be an `EncodingError::unsupported`?

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4985,32 +4986,68 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                    )
+                );
+            },
+            other => unimplemented!("Length operation on unsupported type '{:?}'", other),
```
`EncodingError::unsupported`?

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5524,6 +5561,42 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                    lookup_pure,
+                    vec![
+                        encoded_base_expr,
+                        index.field(vir::Field::new("val_int", vir::Type::Int)),
```
If you need this again, `Encoder::encode_value_field` is slightly better than hard-coding `"val_int"`.

dario23 committed:
Return EncodingError instead of panicking

dario23 committed:
Replace calls to slice::len() with the result of our pure function

i'm not sure if this is the way to go, or rather attach a specification
to the function we actually call (if possible), but this works, at
least. may need some prettyfication/refactoring/simplification.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -2071,6 +2071,74 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+
+                            assert!(args.len() == 1, "unexpected args to slice::len(): {:?}", args);
+                            let slice_operand = self.mir_encoder.encode_operand_expr(&args[0])
+                                .with_span(span)?;
```
Add `.run_if_err(|| cleanup(&self))?` to call the cleanup closure before returning an error. It's an ugly consequence of the hacky support for generics. (Here and for all the `?` below.)

dario23 replied:
moved the whole block into a separate function, so this should only be needed once on the function call.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -2071,6 +2071,74 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+
+                            trace!("{:?} := Slice${:?}$len({:?})", destination, ty, slice_operand);
+
+                            // we need to put a label before, it seems..
```
Why? Have you tried without? The `len` function doesn't reborrow, so I suspect that the `encode_transfer_args_permissions` and the label are not needed.

dario23 replied:
i get a panic saying "thread 'rustc' panicked at 'Location bb0[6] has not yet been encoded', prusti-viper/src/encoder/procedure_encoder.rs:5612:9".

this is with [this testcase](https://github.com/viperproject/prusti-dev/files/6433678/slice-len.rs.txt) and commenting out the label insertion and `encode_transfer_args_permission` call.

[backtrace.txt](https://github.com/viperproject/prusti-dev/files/6433682/backtrace.txt)

i'm not exactly sure, but it seems to me that as this is the last statement in the basic block some permission transfer stuff is being called anyway, to transfer permissions to the next basic block(s), even though we're not doing anything within the statement itself that would require transfering permissions.

dario23 replied:
adding the label, but leaving out the `encode_transfer_args_permissions` call, i get a fold/unfold-error:
```
error: [Prusti internal error] generating fold-unfold Viper statements failed (FailedToObtain(Pred(LabelledOld("l0", Field(Local(_6: Ref(ref$Slice$i32), Position { line: 3, column: 1, id: 73 }), val_ref: Ref(Slice$i32), Position { line: 3, column: 1, id: 73 }), Position { line: 3, column: 1, id: 73 }), read)))
 --> prusti-tests/tests/verify/pass/slices/slice-len.rs:3:1
  |
3 | / fn slice_len_gt_0(x: &[i32]) {
4 | |     assert!(x.len() >= 0);
5 | | }
  | |_^
```

dario23 replied:
[mir.txt](https://github.com/viperproject/prusti-dev/files/6433699/mir.txt)
[slice-len.rs.vir.txt](https://github.com/viperproject/prusti-dev/files/6433702/slice-len.rs.vir.txt)

fpoli replied:
I see, thanks. It seems required to expire the shared reference that is passed as argument to the MIR function call.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -2071,6 +2071,74 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                             }
                         }
 
+                        "core::slice::<impl [T]>::len" => {
```
Could you move the body of this case to a separate method?

dario23 replied:
done.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5245,6 +5360,40 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+    )> {
+        // get rid of ref
+        let slice_ty = match slice_ty.kind() {
+            ty::TyKind::Ref(_, ty, _) => ty,
```
If it's known at the call-site of `encode_slice_types` that `slice_ty` is a `TyKind::Ref`, then I would prefer to remove this match and move it to the caller.

dario23 replied:
ack, moved.

fpoli commented on prusti-viper/src/encoder/mir_encoder/mod.rs:
```diff
@@ -218,22 +218,34 @@ pub trait PlaceEncoder<'v, 'tcx: 'v> {
-                    elem_ty,
-                    None,
-                )
+                    ty::TyKind::Slice(elem_ty) => {
```
Is it possible to add a test case that executes this branch?

dario23 replied:
will do.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4985,32 +5054,78 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                    )?
+                );
+            },
+            ty::TyKind::Slice(..) => {
```
Is it possible to add a test case that executes this branch? I'm not sure how to generate a `mir::Rvalue::Len`. Maybe it's automatically generated by the compiler when it checks the bounds of slice indexes.

dario23 replied:
added a testcase, indeed bounds checks are exercising this case.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4985,32 +5054,78 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                                vir::LocalVar::new(
+                                    String::from("self"),
+                                    vir::Type::TypedRef(ty_name),
+                                ),
```
You use this a lot, so it may be better to define a method such that you can just call `vir::LocalVar::new_typed_ref("self", ty_name)`.

fpoli reviewed (commented):
It looks good. I commented on the diff.

When adding tests, consider adding some also to `verify/fail`. Otherwise, the tests don't fail when we inhale contradictions by mistake.

dario23 committed:
Add vir::LocalVar::new_typed_ref

dario23 committed:
Move slice len impl into separate function

dario23 committed:
Encode havoc and allocation for slice Len mir expr as well

We just use encode_copy_value_assign instead of hand-building
vir::Stmt::Assign statements.

dario23 committed:
Add testcase for slice len, lookup

dario23 committed:
Address some review

dario23 committed:
Use EncodedSliceTypes type to avoid huge tuple

dario23 committed:
Put access permission first in LookupPure encoding

order seems to matter here

dario23 committed:
Add compile-fail test cases

fpoli commented on prusti-tests/tests/verify/fail/arrays-slices/out-of-bounds.rs:
```diff
@@ -0,0 +1,8 @@
+
+
+fn third(s: &[i32]) -> i32 {
+    s[2]  //~ ERROR index out of bounds
```
I would add more text from the error message to check that Prusti says that the statement *may* fail with an index out of bounds.

dario23 replied:
updated the error message.. it's a bit ugly, because i have to include the double quotes now, as the full message is `assertion might fail with ""index out of bounds: the length is {} but the index is {}", move _3, _2"`, but for testing it's clearer now i think.

fpoli replied:
Uh, didn't you fix the "ugliness"? Is the fix outside of this PR?

Aurel300 replied:
Also it's a little strange that the message says "assertion" when there is no assertion as far as the user is concerned.

dario23 replied:
@fpoli what we fixed was a panic/`unreachable!()` when displaying an `AssertKind::BoundsCheck`, the message is now what `fmt_assert_args` prints.
@Aurel300 i think i have to take back what i said in today's meeting; the "assertion might fail with" part comes from [prusti-viper/src/encoder/errors/error_manager.rs:273](https://github.com/viperproject/prusti-dev/blob/master/prusti-viper/src/encoder/errors/error_manager.rs#L273), so we can only really influence what is inside the quotes without extra effort. viper only reports an assertion failure as well, it obviously cannot tell that this is inside a bounds check.

Aurel300 replied:
Ok, I see. However, in the file you link, the `ErrorCtxt::AssertTerminator(...)` is actually something Prusti adds when registering the assertion failure with the error manager. Can we easily (i.e. hopefully without string matching…) detect that a particular assert in MIR corresponds to a bound check and if so, register it with a different `ErrorCtxt`?
This might be a lot of work for a tiny annoyance, but it's worth checking.

fpoli replied:
I think changing "assertion might fail with" to "statement might fail with" would be an improvement for all `AssertTerminator`'s usages.

dario23 replied:
@Aurel300 hm, good idea, added that :-)
@fpoli i guess that general change is something for a different PR

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -2201,6 +2223,70 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        let rhs = vir::Expr::func_app(
+            slice_len_name,
+            vec![
+                slice_operand.clone(),
```
See the Clippy warning(s) reported in Github's diff.

dario23 replied:
adressed, i really like that way of showing the warnings :-)

fpoli reviewed (commented):
It looks good to me, just a few Clippy warnings to fix and a test annotation to change.

dario23 committed:
Avoid unnecessary clones

dario23 committed:
Clarify error message for out-of-bounds slice access

dario23 committed:
Add and use BoundsCheck ErrorCtxt variant

Aurel300 commented on prusti-viper/src/encoder/errors/error_manager.rs:
```diff
@@ -52,6 +52,8 @@ pub enum ErrorCtxt {
     /// A Viper `assert false` that encodes the failure (panic) of an `assert` Rust terminator
     /// Arguments: the message of the Rust assertion
     AssertTerminator(String),
+    /// A Viper `assert false` in the context of a bounds check (if we were able to detect it)
```
> "if we were able to detect it"

Do you know of any cases when the bounds check is not detected? If not, let's be more positive in the comment and leave out the last part.

dario23 replied:
hm, you're right, i don't know of any cases we're missing, as long as rustc has the correct `AssertKind` on them.. removed the last part.

dario23 committed:
Reword

fpoli commented on prusti-viper/src/encoder/errors/error_manager.rs:
```diff
@@ -473,6 +475,13 @@ impl<'tcx> ErrorManager<'tcx>
 
+            ("assert.failed:assertion.false", ErrorCtxt::BoundsCheckAssert) => {
+                PrustiError::verification(
+                    format!("the array or slice index may be out of bounds"),
```
```suggestion
                    "the array or slice index may be out of bounds".to_string(),
```

fpoli committed:
Fix a Clippy warning

dario23 committed:
Remove unused cleanup
